api/opds_for_distributors.py

```diff
@@ -0,0 +1,329 @@
+        },
+        {
+            "key": ExternalIntegration.PASSWORD,
+            "label": _("Library's password or secret key"),
+        }
+    ]
+
+    SUPPORTED_MEDIA_TYPES = [Representation.EPUB_MEDIA_TYPE]
+
+    delivery_mechanism_to_internal_format = {
+        (type, DeliveryMechanism.NO_DRM): type for type in SUPPORTED_MEDIA_TYPES
+    }
+
+    def __init__(self, _db, collection):
+        self.collection_id = collection.id
+        self.username = collection.external_integration.username
+        self.password = collection.external_integration.password
+        self.feed_url = collection.external_account_id
+        self.auth_url = None
+
+    def _request_with_timeout(self, method, url, *args, **kwargs):
+        """Wrapper around HTTP.request_with_timeout to be overridden for tests."""
+        return HTTP.request_with_timeout(method, url, *args, **kwargs)
+
+    def _get_token(self):
+        # If this is the first time we're getting a token, we
+        # need to find the authenticate url in the OPDS
+        # authentication document.
+        if not self.auth_url:
+            response = self._request_with_timeout('GET', self.feed_url)
+
+            if response.status_code != 401:
+                # This feed doesn't require authentication, so
+                # we need to find a link to the authentication document.
+                feed = feedparser.parse(response.content)
+                links = feed.get('feed', {}).get('links', [])
+                auth_doc_links = [l for l in links if l['rel'] == "http://opds-spec.org/auth/document"]
+                if not auth_doc_links:
+                    raise LibraryAuthorizationFailedException()
+                auth_doc_link = auth_doc_links[0].get("href")
+
+                response = self._request_with_timeout('GET', auth_doc_link)
+
+            try:
+                auth_doc = json.loads(response.content)
+            except Exception, e:
+                raise LibraryAuthorizationFailedException()
+            auth_types = auth_doc.get('authentication', [])
+            credentials_types = [t for t in auth_types if t['type'] == "http://opds-spec.org/auth/oauth/client_credentials"]
+            if not credentials_types:
+                raise LibraryAuthorizationFailedException()
+
+            links = credentials_types[0].get('links', [])
+            auth_links = [l for l in links if l.get("rel") == "authenticate"]
+            if not auth_links:
+                raise LibraryAuthorizationFailedException()
+            self.auth_url = auth_links[0].get("href")
+
+        headers = dict()
+        auth_header = "Basic %s" % base64.b64encode("%s:%s" % (self.username, self.password))
+        headers['Authorization'] = auth_header
+        headers['Content-Type'] = "application/x-www-form-urlencoded"
+        token_response = self._request_with_timeout('POST', self.auth_url, data={}, headers=headers)
+        token = json.loads(token_response.content).get("access_token")
```

Comment:
This doesn't need to go in this branch, but we can store the token in a `Credential` rather than fetching a new one every time. This will speed up both metadata imports and fulfillment.